pywr/notebook/__init__.py

```diff
@@ -17,8 +16,110 @@
+
+        It also contains a method to save the node positions of a notebook graph back to a pywr model json file. Note
+        that this method currently odoes not work if the object has be instantiated using a model object. 
+
+        Parameters
+        ----------
+        model : pywr.core.Model or json-dict that describes a model
+            The model to display
+        width : int
+            The width of the svg canvas to draw the graph on
+        height : int
+            The height of the svg canvas to draw the graph on
+        labels : bool
+            If True, each graph node is labelled with its name. If false, the node names are displayed
+            during mouseover events
+        attributes : bool
+            If True, a table of node attributes is displayed during mouseover events
+        css : string
+            Stylesheet data to use instead of default
+        """
+
+        if isinstance(model, Model):
+            self.graph = self.pywr_model_to_d3_json(model, attributes)
+            # TODO update when schema branch is merged
+            self.json = None
+        else:
+            self.graph = self.pywr_json_to_d3_json(model, attributes)
+            with open(model) as d:
+                self.json = json.load(d)
+
+        self.height = height
+        self.width = width
+        self.labels = labels
+        self.attributes = attributes
+
+        if css is None:
+            self.css = draw_graph_css
+        else:
+            self.css = css
+
+    def draw_graph(self):
+        """Draw pywr schematic graph in a jupyter notebook"""
+        js = draw_graph_template.render(
+            graph=self.graph,
+            width=self.width,
+            height=self.height,
+            element="element",
+            labels=self.labels,
+            attributes=self.attributes,
+            css=self.css.replace("\n", "")
+        )
+        display(Javascript(data=js))
+
+    def save_graph(self, filename="model.json"):
+        """Save a copy of the model json with update schematic positions"""
+
+        if self.json is None:
+            warnings.warn("Node positions cannot be saved to json if PywrSchematic object has been instantiated using"
+                          "a pywr model object. Please use a json file path or Python dict instead", stacklevel=2)
+        else:
+            display(Javascript(save_graph_template.render(
+                model_data=json.dumps(self.json),
+                height=self.height,
+                width=self.width,
+                filename=json.dumps(filename)
+            )))
+
+    def to_html(self, filename, title="Model Schematic"):
+        """Save an HTML file of schematic"""
+
+        # TODO add option to get node position from graph that has already been drawn in a notebook
+
+        js = draw_graph_template.render(
+            graph=self.graph,
+            width=self.width,
+            height=self.height,
+            element=json.dumps(".schematic"),
```

Comment:
Why do you need to `json.dumps` this string?

Reply:
Without it, the string does not get properly rendered with `"` in the js file